### PR TITLE
add support for point clouds over flatbuffers

### LINKED ISSUE #88: pointcloud fb interface
For flatbuffers only images are implemented yet. pointclouds are pending.

### PR DESCRIPTION
**Pull request by jarkenau**
This PR adds support for point clouds over Flatbuffers (Closes #88). 

Currently, supported data fields are:
- x,y,z
- r,g,b
- r,gb,a

Left to implement (via separate PRs):

- #188 
- #187

### CODE REVIEW

**Comment by jarkenau** [comment]
I will make it "real" PR after I'm finished with further testing

**Comment by jarkenau** [comment]
@mhoellmann I think it's the easiest way to review if you check if the queries from the python script work as expected. Also, please have a look at `hdf5-fb-pointcloud` and check if you spot possible improvements. The rest of the code should be pretty straight forward.

File: seerep-hdf5/seerep-hdf5-fb/src/hdf5-fb-general.cpp [hunk 1]
```diff
@@ -11,56 +11,53 @@ Hdf5FbGeneral::Hdf5FbGeneral(std::shared_ptr<HighFive::File>& file, std::shared_
 
 void Hdf5FbGeneral::writeAttributeMap(
     const std::shared_ptr<HighFive::DataSet> dataSetPtr,
-    const flatbuffers::Vector<flatbuffers::Offset<seerep::fb::UnionMapEntry>>* attributes)
+    const flatbuffers::Vector<flatbuffers::Offset<seerep::fb::UnionMapEntry>>& attributes)
 {
-  if (attributes)
+  for (auto attribute : attributes)
```
Comment (mhoellmann) [inline seerep-hdf5/seerep-hdf5-fb/src/hdf5-fb-general.cpp:16]: Are you sure that this won't create a segmentation fault, when the attribute map isn't set in the FB msg?

File: seerep-hdf5/seerep-hdf5-fb/src/hdf5-fb-general.cpp [hunk 2]
```diff
@@ -11,56 +11,53 @@ Hdf5FbGeneral::Hdf5FbGeneral(std::shared_ptr<HighFive::File>& file, std::shared_
...
   }
 }
 
 void Hdf5FbGeneral::writeBoundingBoxLabeled(
     const std::string& datatypeGroup, const std::string& uuid,
-    const flatbuffers::Vector<flatbuffers::Offset<seerep::fb::BoundingBoxLabeled>>* boundingboxLabeled)
+    const flatbuffers::Vector<flatbuffers::Offset<seerep::fb::BoundingBoxLabeled>>& boundingboxLabeled)
 {
-  if (boundingboxLabeled && !boundingboxLabeled->size() == 0)
+  if (!boundingboxLabeled.size() == 0)
```
Comment (mhoellmann) [inline seerep-hdf5/seerep-hdf5-fb/src/hdf5-fb-general.cpp:53]: same for this: are you sure that this doesn't create a segmentation fault if the boundingboxLabeled vector in the FB msg isn't set?

File: seerep-hdf5/seerep-hdf5-fb/src/hdf5-fb-general.cpp [hunk 3]
```diff
@@ -89,16 +86,16 @@ void Hdf5FbGeneral::writeBoundingBoxLabeled(
 
 void Hdf5FbGeneral::writeBoundingBox2DLabeled(
     const std::string& datatypeGroup, const std::string& uuid,
-    const flatbuffers::Vector<flatbuffers::Offset<seerep::fb::BoundingBox2DLabeled>>* boundingbox2DLabeled)
+    const flatbuffers::Vector<flatbuffers::Offset<seerep::fb::BoundingBox2DLabeled>>& boundingbox2DLabeled)
 {
   std::string id = datatypeGroup + "/" + uuid;
 
-  if (boundingbox2DLabeled && !boundingbox2DLabeled->size() == 0)
+  if (!boundingbox2DLabeled.size() == 0)
```
Comment (mhoellmann) [inline seerep-hdf5/seerep-hdf5-fb/src/hdf5-fb-general.cpp:93]: same for this: are you sure that this doesn't create a segmentation fault if the boundingbox2DLabeled vector in the FB msg isn't set?

File: seerep-hdf5/seerep-hdf5-fb/src/hdf5-fb-general.cpp [hunk 4]
```diff
@@ -126,15 +123,15 @@ void Hdf5FbGeneral::writeBoundingBox2DLabeled(
 
 void Hdf5FbGeneral::writeLabelsGeneral(
     const std::string& datatypeGroup, const std::string& uuid,
-    const flatbuffers::Vector<flatbuffers::Offset<seerep::fb::LabelWithInstance>>* labelsGeneral)
+    const flatbuffers::Vector<flatbuffers::Offset<seerep::fb::LabelWithInstance>>& labelsGeneral)
 {
   std::string id = datatypeGroup + "/" + uuid;
 
-  if (labelsGeneral && !labelsGeneral->size() == 0)
+  if (labelsGeneral.size() != 0)
```
Comment (mhoellmann) [inline seerep-hdf5/seerep-hdf5-fb/src/hdf5-fb-general.cpp:130]: same for this: are you sure that this doesn't create a segmentation fault if the labelsGeneral vector in the FB msg isn't set?

**Review by mhoellmann — COMMENTED** [review]
This is a really huge PR. For the next feature we should aim for smaller PRs.

I had a quick look at the code without running it. The remarks are just my first thought based purely on the code.

File: seerep-srv/seerep-core-fb/src/core-fb-conversion.cpp [hunk 5]
```diff
@@ -102,6 +102,37 @@ seerep_core_msgs::DatasetIndexable CoreFbConversion::fromFb(const seerep::fb::Po
...
+  seerep_core_msgs::DatasetIndexable dataForIndices;
+  fromFbDataHeader(cloud.header(), dataForIndices.header, seerep_core_msgs::Datatype::PointCloud);
+
+  // set bounding box for images to 0. assume no spatial extent
+  dataForIndices.boundingbox.min_corner().set<0>(0);
+  dataForIndices.boundingbox.min_corner().set<1>(0);
+  dataForIndices.boundingbox.min_corner().set<2>(0);
+  dataForIndices.boundingbox.max_corner().set<0>(0);
+  dataForIndices.boundingbox.max_corner().set<1>(0);
+  dataForIndices.boundingbox.max_corner().set<2>(0);
```
[the remark this answers is not in the document]
Comment (jarkenau) [reply seerep-srv/seerep-core-fb/src/core-fb-conversion.cpp]: It's possible but pretty ugly since we don't wont to iterate over the x,y,z twice. See https://github.com/agri-gaia/seerep/pull/193/commits/3187fcd24bf368e6dbddc64b0165c6e0b8881f05

[the remark this answers is not in the document]
Comment (jarkenau) [reply seerep-srv/seerep-core-fb/src/core-fb-conversion.cpp]: Issue for refactor in https://github.com/agri-gaia/seerep/issues/195

File: seerep-srv/seerep-core-fb/include/seerep-core-fb/core-fb-conversion.h [hunk 6]
```diff
@@ -165,6 +175,14 @@ class CoreFbConversion
   static void
   fromFbDataLabelsGeneral(const flatbuffers::Vector<flatbuffers::Offset<seerep::fb::BoundingBox2DLabeled>>* labelsBB2d,
                           std::vector<seerep_core_msgs::LabelWithInstance>& labelWithInstance);
+  /**
+   * @brief converts the BoundingBoxLabeled with instances of the flatbuffer data message to seerep core specific message
+   * @param labelsBB the BoundingBoxLabeled with instances in the flatbuffer data message
+   * @param labelWithInstance the BoundingBoxLabeled with instances in the data message in seerep core format
+   */
+  static void
+  fromFbDataLabelsGeneral(const flatbuffers::Vector<flatbuffers::Offset<seerep::fb::BoundingBoxLabeled>>* labelsBB,
```
[the remark this answers is not in the document]
Comment (jarkenau) [reply seerep-srv/seerep-core-fb/include/seerep-core-fb/core-fb-conversion.h:184]: I don't think we need to rename it, since they serve the same purpose and are just overloaded with different parameters.

File: seerep-hdf5/seerep-hdf5-fb/src/hdf5-fb-general.cpp (hunk 2 above)
Comment (jarkenau) [reply seerep-hdf5/seerep-hdf5-fb/src/hdf5-fb-general.cpp:53]: Before calling this function, I checked it the field is set, but it still makes sense to add it here for additional safety.

Comment (jarkenau) [reply seerep-hdf5/seerep-hdf5-fb/src/hdf5-fb-general.cpp:53]: It will cause a segmentation fault, but generally all the functions use references and to call the function you would have dereference a `null pointer` which is undefined behavior. Now the question is, how we want to handle that, let it crash, to force the null checks' before the function is called or also catch it in the function.

Comment (mhoellmann) [reply seerep-hdf5/seerep-hdf5-fb/src/hdf5-fb-general.cpp:53]: exactly. For that reason the pointer was the argument for the methode previously. With this the pointer could be a `null pointer` and the check was done in the methode. Otherwise we'd have to check if the funtion is not `null` before each call. 

We cannot avoid the `null pointers`, because the sub messages of the fb message are optional and the pointer to the submessage is `null` if the submessage is not used.

In summary: we have to check either before each methode call, which may be forgotten and create a crash of the server or we check it just once inside of the methode. I'd prefer inside of the methode

File: seerep-srv/seerep-core-fb/include/seerep-core-fb/core-fb-conversion.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (mhoellmann) [reply seerep-srv/seerep-core-fb/include/seerep-core-fb/core-fb-conversion.h:184]: Well, yes, but not quite. The naming convention in that file is "fromFbData" + whatever part of the data-msg is converted. e.g. `fromFbDataHeader` for the conversion of the Header sub-msgs of the fb data msg.

I think I made the first error by naming the second methode also `fromFbDataLabelsGeneral` instead of `fromFbDataLabelsBb2d` as it would have been consistently.

The third methode should be named `fromFbDataLabelsBb`

File: seerep-hdf5/seerep-hdf5-fb/include/seerep-hdf5-fb/hdf5-fb-pointcloud.h [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+
+  /**
+   * @brief Writes information about point fields as attributes to hdf5
+   *
+   * @param object HighFive object to write to (representing a data set or data group)
+   * @param pointFields flatbuffers vector of the point fields to write
+   */
+  template <typename T>
+  void writePointFieldAttributes(HighFive::AnnotateTraits<T>& object,
+                                 const flatbuffers::Vector<flatbuffers::Offset<seerep::fb::PointField>>& pointFields)
```
Comment (mhoellmann) [inline seerep-hdf5/seerep-hdf5-fb/include/seerep-hdf5-fb/hdf5-fb-pointcloud.h:77]: Here also pointer and check if pointer not null. Should happen, because then it's no pointcloud at all, but anyway

File: seerep-hdf5/seerep-hdf5-fb/src/hdf5-fb-pointcloud.cpp [hunk 8]
```diff
@@ -0,0 +1,814 @@
...
+namespace seerep_hdf5_fb
+{
+Hdf5FbPointCloud::Hdf5FbPointCloud(std::shared_ptr<HighFive::File>& file, std::shared_ptr<std::mutex>& writeMtx)
+  : Hdf5FbGeneral(file, writeMtx)
+{
+}
+
+// TODO refactor again, still not happy with the API
+
+void Hdf5FbPointCloud::writePointCloud2(const std::string& id, const seerep::fb::PointCloud2& cloud,
```
Comment (mhoellmann) [inline seerep-hdf5/seerep-hdf5-fb/src/hdf5-fb-pointcloud.cpp:14]: Methodes are too long in this file. Please break them down in multiple methodes.

Comment (jarkenau) [reply seerep-hdf5/seerep-hdf5-fb/src/hdf5-fb-pointcloud.cpp:12]: The methods should now have a reasonable length, except:

https://github.com/agri-gaia/seerep/blob/4524908f6db045517f4509cbca945f0e7793d217/seerep-hdf5/seerep-hdf5-fb/src/hdf5-fb-pointcloud.cpp#L201-L320

But it's hard to break the method down without using >8 parameters, is it sill  worth it?

File: seerep-hdf5/seerep-hdf5-fb/include/seerep-hdf5-fb/hdf5-fb-pointcloud.h [hunk 9]
```diff
@@ -0,0 +1,325 @@
...
+   * @brief struct to store information about the PointCloud2
+   *
+   */
+  struct CloudInfo
+  {
+    bool has_points = false;
+    bool has_rgb = false;
+    bool has_rgba = false;
+    bool has_normals = false;
+  };
```
Comment (jarkenau) [inline seerep-hdf5/seerep-hdf5-fb/include/seerep-hdf5-fb/hdf5-fb-pointcloud.h:67]:
```suggestion
  struct CloudInfo
  {
    bool has_points = false;
    bool has_rgb = false;
    bool has_rgba = false;
    bool has_normals = false;
    uint32_t height;
    uint32_t width;
    ....
  };
```

File: seerep-hdf5/seerep-hdf5-fb/include/seerep-hdf5-fb/hdf5-fb-pointcloud.h [hunk 10]
```diff
@@ -0,0 +1,325 @@
...
+   * @param pointStep the point step of the point cloud
+   * @param height the height of the point cloud (1 of unorganized point clouds)
+   * @param width the width of the point cloud
+   * @param groupPtr shared pointer to the data group of the point cloud.
+   *        (Used to add a bounding box of x,y,z as an attribute of the point cloud)
+   * @param boundingBox reference to vector to write the computed boundingBox
+   */
+  void writePoints(const std::string& id, const std::vector<uint32_t>& offsets, const uint8_t* data, uint32_t pointStep,
+                   uint32_t height, uint32_t width, const std::shared_ptr<HighFive::Group>& groupPtr,
+                   std::vector<float>& boundingBox);
```
Comment (jarkenau) [inline seerep-hdf5/seerep-hdf5-fb/include/seerep-hdf5-fb/hdf5-fb-pointcloud.h:130]:
```suggestion
  void writePoints(const std::string& id, const uint8_t* data, PointCloudInfo& info, const std::shared_ptr<HighFive::Group>& groupPtr,
                   std::vector<float>& boundingBox);
```

**Review by jarkenau — COMMENTED** [review]
One way that I could improve the API of the `Hdf5FbPointCloud` Class would be to extend the `CloudInfo` struct with more information like height, width etc. and then use it in function parameters. But it's just another abstraction.

File: seerep-hdf5/seerep-hdf5-fb/include/seerep-hdf5-fb/hdf5-fb-pointcloud.h (hunk 9 above)
Comment (mhoellmann) [reply seerep-hdf5/seerep-hdf5-fb/include/seerep-hdf5-fb/hdf5-fb-pointcloud.h:67]: I think it's fine for now. We can rethink it when needed.

**Review by mhoellmann — APPROVED** [review]
